torch/_cuda_sanitizer/_sanitizer.py

```diff
@@ -0,0 +1,490 @@
+from typing import Any, Dict, Iterator, List, Optional, Set, Tuple, TypeVar
+
+import torch
+import torch.utils._cuda_trace as cuda_trace
+from torch.utils._cuda_trace import _CUDA_HOOK_TRACE_OFFSET
+from torch.utils._python_dispatch import (
+    TorchDispatchMode,
+    _PYTHON_DISPATCH_TRACE_OFFSET,
+)
+from torch.utils._pytree import tree_map
+
+
+TK = TypeVar("TK")
+TVa = TypeVar("TVa")
+TVb = TypeVar("TVb")
+
+DataPtr = int
+StreamId = int
+EventId = int
+SeqNum = int
+
+logger = logging.getLogger(__name__)
+
+
+class AccessType(enum.Enum):
+    READ = enum.auto()
+    WRITE = enum.auto()
+
+    def __str__(self):
+        return "reading from" if self is AccessType.READ else "writing to"
+
+
+@dataclass
+class Access:
+    type: AccessType
+    seq_num: SeqNum
+    stream: StreamId
+    operator: str
+    names: List[str]
+    stack_trace: List[str]
+
+
+class SynchronizationError(Exception):
+    """Base class for errors detected by CSAN."""
+
+    pass
+
+
+class UnsynchronizedAccessError(SynchronizationError):
+    def __init__(
+        self,
+        data_ptr: DataPtr,
+        allocation_trace: Optional[List[str]],
+        current_access: Access,
+        previous_access: Access,
+    ):
+        with io.StringIO() as message:
+            message.write(
+                textwrap.dedent(
+                    f"""\
+                    ============================
+                    CSAN detected a possible data race on tensor with data pointer {data_ptr}
+                    Access by stream {current_access.stream} during kernel:
+                    {current_access.operator}
+                    {current_access.type} argument: {', '.join(current_access.names)}
+                    With stack trace:
+                    """
+                )
+            )
+            message.write(f"{''.join(current_access.stack_trace)}\n")
+            message.write(
+                textwrap.dedent(
+                    f"""\
+                    Previous access by stream {previous_access.stream} during kernel:
+                    {previous_access.operator}
+                    {previous_access.type} argument: {', '.join(previous_access.names)}
+                    With stack trace:
+                    """
+                )
+            )
+            message.write(f"{''.join(previous_access.stack_trace)}\n")
+            if allocation_trace:
+                message.write(
+                    "Tensor was allocated with stack trace:\n"
+                    f"{''.join(allocation_trace)}"
+                )
+            else:
+                message.write("Trace for tensor allocation not found.")
+            super().__init__(message.getvalue())
+
+
+def format_log_message(message: str) -> str:
+    return " ".join(line.strip() for line in message.strip().splitlines())
+
+
+@dataclass
+class TensorInfo:
+    allocation_trace: Optional[List[str]]
+    reads: List[Access] = field(default_factory=list)
+    write: Optional[Access] = None
```

Comment:
A comment for this class would help explicate why there can be multiple reads but only one write